pkg/apis/gcp/validation/infrastructure.go

```diff
@@ -125,13 +125,56 @@ func ValidateInfrastructureConfig(infra *apisgcp.InfrastructureConfig, nodesCIDR
 		}
 	}
 
-	if infra.Networks.CloudNAT != nil && infra.Networks.CloudNAT.NatIPNames != nil && len(infra.Networks.CloudNAT.NatIPNames) == 0 {
-		allErrs = append(allErrs, field.Invalid(networksPath.Child("cloudNAT", "natIPNames"), infra.Networks.CloudNAT.NatIPNames, "nat IP names cannot be empty"))
+	if infra.Networks.CloudNAT != nil {
+		allErrs = append(allErrs, ValidateCloudNatConfig(infra.Networks.CloudNAT, networksPath)...)
 	}
 
 	return allErrs
 }
 
+// ValidateCloudNatConfig validates the config of the CloudNat. We intentionally keep the validation light, only
+// checking for gotchas (e.g. the port counts having to be powers of two) and obvious errors.
+func ValidateCloudNatConfig(config *apisgcp.CloudNAT, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	cloudNatPath := fldPath.Child("cloudNAT")
+
+	if config != nil && config.NatIPNames != nil && len(config.NatIPNames) == 0 {
+		allErrs = append(allErrs, field.Invalid(cloudNatPath.Child("natIPNames"), config.NatIPNames, "nat IP names cannot be empty."))
+	}
+
+	if config.EndpointIndependentMapping != nil && config.EnableDynamicPortAllocation && config.EndpointIndependentMapping.Enabled {
+		// There is no more fitting field.Error (e.g. field.MutuallyExclusive) so we put the blame on one flag and use the error msg
+		allErrs = append(allErrs, field.Invalid(cloudNatPath.Child("enableDynamicPortAllocation"), config.EnableDynamicPortAllocation, "dynamic port allocation may not be enabled at the same time as endpoint independent mapping."))
+	}
+
+	if config.EnableDynamicPortAllocation {
+		if config.MaxPortsPerVM != nil && !isPowerOfTwo(*config.MaxPortsPerVM) {
+			allErrs = append(allErrs, field.Invalid(cloudNatPath.Child("maxPortsPerVM"), config.MaxPortsPerVM, "maxPortsPerVM must be a power of two."))
+		}
+		if config.MinPortsPerVM != nil && !isPowerOfTwo(*config.MinPortsPerVM) {
+			allErrs = append(allErrs, field.Invalid(cloudNatPath.Child("minPortsPerVM"), config.MinPortsPerVM, "minPortsPerVM must be a power of two if dynamic port allocation is enabled."))
```

Comment:
This check is relevant even outside the DPA section right ?

Reply:
No, if DPA is active min ports _must_ be a power of two and _at least_ 32. If it is not active, it may be set to a value as low as 2 and there is no additional requirement.